Source/Core/Core/PowerPC/Expression.cpp

```diff
@@ -0,0 +1,241 @@
+static double CastFunc(expr_func* f, vec_expr_t* args, void* c)
+{
+  if (vec_len(args) != 1)
+    return 0;
+  return Common::BitCast<T>(static_cast<U>(expr_eval(&vec_nth(args, 0))));
+}
+
+static std::array<expr_func, 21> g_expr_funcs{{
+    // For internal storage and comparisons, everything is auto-converted to Double.
+    // If u64 ints are added, this could produce incorrect results.
+    {"read_u8", HostReadFunc<u8>},
+    {"read_s8", HostReadFunc<s8, u8>},
+    {"read_u16", HostReadFunc<u16>},
+    {"read_s16", HostReadFunc<s16, u16>},
+    {"read_u32", HostReadFunc<u32>},
+    {"read_s32", HostReadFunc<s32, u32>},
+    {"read_f32", HostReadFunc<float, u32>},
+    {"read_f64", HostReadFunc<double, u64>},
+    {"write_u8", HostWriteFunc<u8>},
+    {"write_u16", HostWriteFunc<u16>},
+    {"write_u32", HostWriteFunc<u32>},
+    {"write_f32", HostWriteFunc<float, u32>},
+    {"write_f64", HostWriteFunc<double, u64>},
+    {"u8", CastFunc<u8>},
+    {"s8", CastFunc<s8, u8>},
+    {"u16", CastFunc<u16>},
+    {"s16", CastFunc<s16, u16>},
+    {"u32", CastFunc<u32>},
+    {"s32", CastFunc<s32, u32>},
+    {},
+}};
+
+void ExprDeleter::operator()(expr* expression) const
+{
+  expr_destroy(expression, nullptr);
+}
+
+void ExprVarListDeleter::operator()(expr_var_list* vars) const
+{
+  // Free list elements
+  expr_destroy(nullptr, vars);
+  // Free list object
+  delete vars;
+}
+
+Expression::Expression(std::string_view text, ExprPointer ex, ExprVarListPointer vars)
+    : m_text(text), m_expr(std::move(ex)), m_vars(std::move(vars))
+{
+  for (auto* v = m_vars->head; v != nullptr; v = v->next)
+  {
+    const std::string_view name = v->name;
+    VarBinding bind;
+
+    if (name.length() >= 2 && name.length() <= 3)
+    {
+      if (name[0] == 'r' || name[0] == 'f')
+      {
+        char* end = nullptr;
+        const int index = std::strtol(name.data() + 1, &end, 10);
+        if (index >= 0 && index <= 31 && end == name.data() + name.length())
+        {
+          bind.type = name[0] == 'r' ? VarBindingType::GPR : VarBindingType::FPR;
+          bind.index = index;
+        }
+      }
+      else if (name == "lr")
+      {
+        bind.type = VarBindingType::SPR;
+        bind.index = SPR_LR;
+      }
+      else if (name == "ctr")
+      {
+        bind.type = VarBindingType::SPR;
+        bind.index = SPR_CTR;
+      }
+      else if (name == "pc")
+      {
+        bind.type = VarBindingType::PCtr;
+      }
+    }
+
+    m_binds.emplace_back(bind);
+  }
+}
+
+std::optional<Expression> Expression::TryParse(std::string_view text)
+{
+  ExprVarListPointer vars{new expr_var_list{}};
+  ExprPointer ex{expr_create(text.data(), text.length(), vars.get(), g_expr_funcs.data())};
+  if (!ex)
+    return std::nullopt;
+
+  return Expression{text, std::move(ex), std::move(vars)};
+}
+
+double Expression::Evaluate() const
+{
+  SynchronizeBindings(SynchronizeDirection::From);
+
+  double result = expr_eval(m_expr.get());
```

Comment:
I think most calls of `expr_eval` should check if they return `NaN` (or warn the user) as this can propagate weird behaviour such as `x != x` being true and `NaN` could also disappear in some operations (i.e. pow(NAN,0) -> 1).

I managed to replicate a similar "misbehaviour" _(though, that's the one to expect from NaN)_ by doing a condition similar to this:
```C++
// if r0 is zero, my_var == my_var is false
my_var = r0 / r0, my_var == my_var
```

Reply:
Should NaN just return true/break? If not how should it come out in the BP check?

For warning the user, should the log be used?  I'm not sure popups are easy to implement in this area.

Reply:
Assuming a user might want to check for NaN in some circumstances, I guess leaving the behaviour as is would be fine.

I think warning the user via a log message and OSD message is the way to go. Some parts of the debugger (ex: StepOut button) are already using OSD messages.

Reply:
Ok, a little stuck here. I figured out the OSD and the log.  I can send it if the return value is NaN, but have no idea on ` x = 0/0, x == x` as the return value is false, not NaN.  Messing with intermittent logic requires changing expr.h.

If a NaN value pops up at any time, should expr.h just stop and error out, returning NaN? Not sure how hard that is it do.

Reply:
If there is a NaN at some point I think you should let it proceed normally. However, I do think it's important to warn the user when such case happens because I don't think most users are expecting it.

I think I found some flaws in the code that might cause a possible undefined behaviour or segfault. I'm going to investigate it further to make sure I'm not wrong.